packages/composites/gm-token/src/transport/price.ts

```diff
@@ -0,0 +1,287 @@
+            asks: [...(priceData[asset]?.asks || []), ask],
+          }
+
+          priceProviders[base] = priceProviders[base]
+            ? [...new Set([...priceProviders[base], source.name])]
+            : [source.name]
+        } catch (error) {
+          const e = error as Error
+          logger.error(
+            `Error fetching data for ${asset} from ${source.name}, url - ${source.url}: ${e.message}`,
+          )
+        }
+      })
+
+      promises.push(...assetPromises)
+    }
+
+    await Promise.all(promises)
+
+    this.validateRequiredResponses(priceProviders, sources, assets, dataRequestedTimestamp)
+
+    const medianValues = this.calculateMedian(assets, priceData)
+
+    const prices = medianValues.map((v) => ({
+      ...v,
+      ask: toFixed(v.ask, decimals[v.asset as keyof typeof decimals]),
+      bid: toFixed(v.bid, decimals[v.asset as keyof typeof decimals]),
+    }))
+
+    return {
+      prices,
+      sources: priceProviders,
+    }
+  }
+
+  private calculateMedian(assets: string[], priceData: PriceData) {
+    return assets.map((asset) => {
+      // Since most of the gm markets have the same long and index tokens, we need to remove duplicate values from duplicate requests
+      const medianBid = median([...new Set(priceData[asset].bids)])
+      const medianAsk = median([...new Set(priceData[asset].asks)])
+      return { asset, bid: medianBid, ask: medianAsk }
+    })
+  }
+
+  private unwrapAsset(asset: string) {
+    if (asset === 'WBTC.b') {
+      return 'BTC'
+    }
+    if (asset === 'WETH') {
+      return 'ETH'
+    }
+    return asset
+  }
+
+  /*
+    For every asset check that we received responses from the required number of source EAs to accurately calculate the median price of the asset.
+  */
+  private validateRequiredResponses(
+    priceProviders: Record<string, string[]> = {},
+    sources: Source[],
+    assets: string[],
+    dataRequestedTimestamp: number,
+  ) {
+    const allSource = sources.map((s) => s.name)
+    if (!Object.entries(priceProviders)?.length) {
+      throw new AdapterDataProviderError(
+        {
+          statusCode: 502,
+          message: `Missing responses from '${allSource.join(',')}' for all assets.`,
+        },
+        {
+          providerDataRequestedUnixMs: dataRequestedTimestamp,
+          providerDataReceivedUnixMs: Date.now(),
+          providerIndicatedTimeUnixMs: undefined,
+        },
+      )
+    }
+
+    assets.forEach((asset) => {
+      const base = this.unwrapAsset(asset)
+      const respondedSources = priceProviders[base]
+
+      if (respondedSources.length < this.settings.MIN_REQUIRED_SOURCE_SUCCESS) {
+        const missingSources = allSource.filter((s) => !respondedSources.includes(s))
+        throw new AdapterDataProviderError(
+          {
+            statusCode: 502,
+            message: `Cannot calculate median price for '${asset}'. Missing responses from '${missingSources.join(
```

Comment:
Can we include the `MIN_REQUIRED_SOURCE_SUCCESS` in this error message? Will help with debugging, as will show if it is using the default or has been configured to a different value. 

```suggestion
            message: `Cannot calculate median price for '${asset} - require ${MIN_REQUIRED_SOURCE_SUCCESS} responses from source EA's but only received ${respondedSources.length}'. Missing responses from '${missingSources.join(
```

Reply:
@alecgard i used a little bit more descriptive message. let me know if you still prefer the one that you suggested

Reply:
Nope yours looks great!